Listener to set onCampus filter in PrimoCentral context senitively

olli-gold opened the pull request:
This PR improves PrimoCentral results.

The current design has a mechanism to autodetect the correct PrimoCentral context by the IP of the VuFind-user. This means, that a person from outside of this IP range cannot get PrimoCentral results.

We want to make PrimoCentral available for everyone using our TUBfind, if we are allowed to do that. PrimoCentral has the parameter onCampus in its API to indicate, if the user is on campus (and might see some restricted results) or not. Currently this filter is always set to true. This is okay, as long as the user's IP address is matching one of the Institution->regex-settings. But if the user matches the Catch-All-rule, this means, that he/she is NOT on campus.

This PR deals with that, setting the onCampus parameter context sensitively for the user's IP address.

I have also prepared a test for the new listener. But as I have included some additional logic into the PrimoBackendFactory, should this perhaps also get tested?

demiankatz commented:
Thanks for sharing this! A few initial comments, based on a fairly cursory review:

1.) Travis reports a handful of style issues that we should deal with... but nothing too difficult, it would appear.

2.) Generally speaking, I'm not too worried about test coverage for factories -- as glue code, they are difficult to test because they pull things from so many different places, and since they're mostly just assembling existing components, they don't tend to offer a lot of interesting logic to test. However, thinking that way only works if the factories are kept as light-weight as possible, with any really complex logic moved into individual components. In the case of your changes to PrimoBackendFactory, some of the logic is fairly complex and might merit testing -- but testing it as part of the factory will probably be a bit of a hassle, requiring you to set up a lot of mock objects. This might be a sign that it ought to move elsewhere, though I'm not sure exactly where, and creating a separate class just for dealing with this functionality is probably overkill (though it would certainly make testing easier). Worth giving a little more thought too, but I'm not overly concerned one way or another.

3.) I notice that you have a lot of methods that refer to "onCampusRange" but I think "onCampusPermission" is probably a more accurate name. I'm guessing that perhaps this began life as a separate range setting and later evolved into a permission-based mechanism.

4.) This also makes me wonder if we need a way of using the new permission system to replace the current [Institutions] section of Primo.ini. An [InstitutionPermissions] section that maps permission name to institution code might make for a more readable configuration, and I don't think it would be too hard to write code to map the legacy configuration to a new Permissions-based version for backward compatibility. If you agree, perhaps this would be a worthwhile pull request as a precursor to merging this one (though, of course, I don't mean to make more work for you if you don't think this would be valuable).

5.) Speaking of permissions.ini, it might be worthwhile to define a standard name for the new permission and add documentation about it (perhaps with a commented-out example configuration) in permissions.ini to make the configuration easier for people who need it.

What do you think?

olli-gold commented:
Thank you for the comments, Demian!

Perhaps we could move all of the logic into the listener. I think it makes much sense to replace the [Institutions] section, and this could also get moved into the onCampusListener. The name still fits, even if the functionality is extended, and this could also get tested much easier.

I'd love to do that (saving you some work) and I can try to continue tomorrow (or next week as I'm out of office on Thursday and Friday) with that. As I mentioned in the mailinglist, there's no need to hurry and perhaps some others also have ideas about that.

demiankatz commented:
Thanks, that sounds like a good plan. I'll mention this on the dev call today, but I don't know if there will be any other Primo users present. As you say, we can take our time and solicit wider feedback as we go.

olli-gold commented:
I have completely rewritten the implementation and have also addressed moving the settings to the new permission system. The test class for the onCampusListener is already adjusted, but I want to add another test class for the new PrimoPermissionController. I will do that next week, in the meantime any comment on this is appreciated!

demiankatz commented:
Thanks for the updates!  A few comments:

1.) I'd rename PrimoPermissionController to something else, like PrimoPermissionManager or PrimoPermissionHandler. Since "controller" has a specific meaning in Zend Framework MVC, I think using that word in the name might cause confusion.

2.) Do we need to have the separate code and permissionRule settings in the config? What if we use permissionRule as an array, so instead of:

<pre>
code[] = "MYINSTITUTION"
permissionRule[] = "primoPermission.MEMBER1"
</pre>

we instead do:

<pre>
permissionRule[MYINSTITUTION] = "primoPermission.MEMBER1"
</pre>

I think that might make this easier to parse and understand, provided the institution code will never contain weird characters that mess up the .ini parsing.

3.) Does this change eliminate support for the old [Institutions] section entirely? If so, we should at very least detect if that section is defined and throw an exception noting that configurations need to be updated. If we're feeling more ambitious, we could add upgrade logic to VuFInd\Config\Upgrade, though given the small user base of Primo Central in VuFind, that's quite possibly more work than it's worth.

4.) Looks like this needs a run of php-cs-fixer to clean up some things (plus shortening of one long line). Let me know if you need help with that.

I'll probably have more comments if I take a deeper look at the code, but hopefully this at least helps get things moving along! I'll try to spend more time looking at the next revision, so please let me know when you're ready for another review.

- Demian

olli-gold commented:
Demian, thanks for the new suggestions. 

1. I have renamed the class and have chosen PrimoPermissionHandler as the new name.
2. Great idea! I have changed that.
3. I have eliminated the section from the config file, but support for the old settings should still work. I had removed it to make it clearer that these settings are deprecated. I have changed that again and have restored the old settings, and just added a comment recommending to use the new settings.
4. I will take care of the style issues again...

Thank you again, and I guess you can make a new review.

olli-gold commented:
I think I have fixed the style issues, but I do not understand, what php-cs-fixer means on number 3 on the check of 489c550. So I guess, this is not fixed.

demiankatz commented:
Thanks for the latest progress! Some new comments:

1.) The php-cs-fixer issues had to do with old array syntax (array('a' => 'b') instead of ['a' => 'b']) and extra whitespace on the ends of lines. I've run `phing php-cs-fixer` and committed the results to clean that up.

2.) If you prefer to remove the [Institutions] section from Primo.ini, that's okay with me -- I wasn't requesting its restoration so much as checking that we addressed its deprecation somewhere. It might be better to take it out again, but throw some kind of deprecation notice in the area where it is parsed. I don't think we have a large enough installed base of Primo users that it's worth making the code extra-complex just for the sake of backward compatibility. I just want to be sure that if something old stops working, there's an obvious red flag indicating how to fix it (or, if easy, some kind of auto-migration).

3.) In InjectOnCampusListener you have a getOnCampus() function that returns no value. I'd recommend changing this to a real getter for clarity -- i.e. initialize isOnCampus to null in the constructor, then if it's null in the getter, execute the logic to resolve it to true or false, and then return the result.

4.) I've revised the comments in Primo.ini in an effort to add clarity (and to fix a couple of typos). Please review and make further adjustments if I've said anything inaccurate. I think we could use further clarification on the behavior of defaultPermissionRule. I'm a bit confused by the current comments, but I'm not sure if this is because the comments need to be made more clear or because that setting belongs in a different area. Are you saying that if defaultPermissionRule is unmatched, we don't set an institution code at all? If so, I think that should be made more clear in the comment. Or is the idea that the default code is always used, but defaultPermissionRule controls some other parameter? If that's the case, I'm wondering if that belongs somewhere other than in the section for determining institution codes (or maybe the section should be renamed to something broader, like simply [Permissions]).

olli-gold commented:
Demian,

thank you again for the new comments and cleaning up the php-cs-fixer issue.

2.) Thinking about that, I would leave the [Institutions] section in Primo.ini as long as it is supported. As I said, it is still supported in the code, but I would recommend using the new section. So its probably the best solution to leave it in Primo.ini as long as the functions to support it are not dropped.

3.) I have renamed the getOnCampus method, as this is not exactly a getter, but more a setter for the onCampus attribute (or a checker for this attribute). So I have renamed it to checkOnCampus. Is that okay?

4.) Perhaps this really was a little bit unclear. So I'll explain more detailed here and improve the comments. 
The VuFind admin has two options to use the PrimoPermission system:
* with the permissionRules (for every institution in a consortium one line)
* with the default permissionRule and the default code. 
If none of the permissionRules is granted (or if none is configured), the default code is used for PrimoCentral. But than we have still to decide, what the user should see. Thats where we need the defaultPermissionRule - if this is granted, the user is on campus and can see all results. If not, he should only see the open results, still using the defaultCode.
If no default code is set, this means, that a user, who is not matching any of the permissionRules, has no access to PrimoCentral at all. A defaultPermissionRule is not necessary in this case, because it has no code associated with it.
One example: We have only one PrimoCentral code, we can use the defaultCode and defaultPermissionRule, if we want all users to be allowed to use the PrimoTab. If we don't want that, we can use the permissionRule['<ourCode>'] = ourMemberPermissionRule, where ourMemberPermissionRule is a permissionRule in permissions.ini.

I hope, this clarifies logic a bit. I don't think another section is necessary, but the comments should be better... I tried to improve them, and I hope they are clear enough now.

olli-gold commented:
I have tested my code again with different configurations and found one thing, that could get improved.
If the configuration is as follows:
permissionRule['MYINSTITUTION'] = "primoPermission.onCampus"
and primoPermission.onCampus does not exist, the user gets no results at all. This is correct on the one hand, but perhaps it would be better to see a note telling the user, that he is not allowed to see anything and what he could do against that (Tell the library about it or ask them).
If MYINSTITUTION did not exist, its the same (no results, but also no exceptions or notes). I guess we could need another level of error management in these cases.

demiankatz commented:
To reply to various points above:

2.) If you fill in both [Institutions] and [InstitutionPermission] sections, how do they interact? Does one trump the other? If we keep them both, we should document how they interact... and if we're planning to eventually eliminate [Institutions], I think adding a deprecation message would be wise.

3.) checkOnCampus is definitely better than getOnCampus, though you could easily make it into a true getter by adding `return $this->onCampus;` at the end of the function, and then you could use the call to getOnCampus directly instead of accessing the class property. That's probably a bit more clear, unless there's some particular reason you prefer direct property access instead of the use of a getter.

4.) This is becoming more clear, but I'm still confused about why we only need the permission rule for the default code. It sounds like we're trying to determine two things: the institution code to use, and whether or not to display full results. Isn't it possible that a user might want to determine these two things independently? For example, a location code might be tied to a user attribute, while permission might be tied to an IP range. So my question is, do we need a mechanism for providing two permission rules for each location, or can we have a single permission rule that is combined with all locations? Or am I totally confused, and it's not actually this complicated?

5.) Regarding the issues in your second comment, I think we should throw an exception in any scenario where there is clearly a configuration problem (for example, if a permission is completely undefined, that's probably an error). However, as you say, there may be some scenarios where a user simply does not have access to any results. In this case, it may make sense to build a recommendation module that could be loaded as a default_noresults_recommend[] in Primo.ini to display suggestions. The challenge is figuring out how to differentiate between a search that yields zero results and a permission issue. (Perhaps the module could trigger a search for some very common word, like 'book', and only display permission messages if that search also comes back empty). In any case, I'd recommend putting in exception throwing now if there's an easy way to detect problem situations (there might not be -- I can't remember how easy it is to detect undefined permissions) and perhaps addressing the recommendation module as a separate project (it might be interesting to try to build this as a generic tool that could be repurposed for other search backends as well -- i.e. make it configurable to accept a backend name and a search query so you could run different tests in different contexts).

As always, thanks for your progress on this!

olli-gold commented:
Thanks for the new comment.

About 2.): The [InstitutionPermission] is used, when its configured, though it trumps the old [Institution] section. If someone of the Primo users is upgrading and has only the [Institution] section, it will get used. If he also upgraded the config file and has configured the new section, this one is getting used. Thats why I added a depraction note to the config file. But (as always) you are right: this depraction message could also get logged to inform the VuFind administrator, that something has changed and he should upgrade the config file as well.

About 3.): Well, if you recommend working with the getter method rather than with the property directly, I will do that. I have changed the code again ;-)

About 4.): A PrimoCentral code is never tied toa user attribute, normally one institution has one code. But for consortial environments it is possible that different codes need to be managed. 
I'll supply another example to make it clearer.
This is my sample configuration:
permissionRule['MYINST'] = "primo.MYINST"
permissionRule['YOURINST'] = "primo.YOURINST"
defaultCode = "DEFAULT"
defaultPermissionRule = "primo.DEFAULT"

So, we have a consortium here: MYINST and YOURINST. If a user is not part of one of the institutions, he will fall back to the DEFAULT code.

Assuming I am a user coming from an IP address, which is matching the permissionRule primo.MYINST (as I am member of MYINST). VuFind will use the code MYINST to communicate with PrimoCentral. PrimoCentral knows now, that I am authorized to see all of the results (because the code was determined by matching my IP address). 

Now I am a user, who is coming from an IP address, which is not matching any of the permissionRules. VuFind will now use the code DEFAULT to communicate with PrimoCentral. Now PrimoCentral needs to know, if I am authorized to see all of the results (this is what the onCampus-parameter is doing). As the defaultPermissionRule does not match, PrimoCentral only shows the free or open results. But the defaultCode is still applied.

All right? 

About 5.): I will try to find a way to throw exceptions, when the configuration uses a permissionRule, which is not existing. And I will also try to take care of what happens, when the user is not allowed to see results because of a permission issue. Then, I guess, this can be completed and we can address the recommendation module later, as you also suggested.

demiankatz commented:
Great, looks like good progress is being made here!

About 4.) Imagine this hypothetical scenario: there is a consortial VuFind instance, where users can log in either on or off campus. However, each user account has an attribute which indicates which consortial library is the user's home. Thus, we may be able to identify a different institution code for each user whether they are on or off campus. It seems to me that it's possible that freely visible results could still vary from institution to institution, and thus we might still need to account for both situations. However, if the freely visible results are the same regardless of institution code, then I would agree that the current configuration covers all of the cases.

About 5.) Sounds good. Please let me know if I can help in any way -- if you can't figure out how to do something, feel free to ask before you waste too much of your own time. If I jog my memory a bit, I may be able to find the right path a little quicker since I have at least seen all of the code before.

olli-gold commented:
"However, each user account has an attribute which indicates which consortial library is the user's home. Thus, we may be able to identify a different institution code for each user whether they are on or off campus." - Yes, right, we cannot determine right now, which library is the users home library, when he logs in. The only way to do that is to select the home library using the IP address. This might be a usefull addition, right...
I will think about that again, tomorrow.

demiankatz commented:
Using the permission system, you could determine the home library based on the user account with the existing code -- it would just require the permissions to be configured appropriately (for example, using some of the Shibboleth settings, or a custom handler for some other permission system). So I don't think you need to make changes to the code to allow user-based code determination... but my point is that there may be separate criteria for figuring out institution code vs. figuring out level of access.

olli-gold commented:
"but my point is that there may be separate criteria for figuring out institution code vs. figuring out level of access." -- OK, now I think I understand your point. As I mentioned earlier, we have only one institution code. This institution code applies for all of the TUBfind users and we can offer it to anyone. But we have to be sure, that we are setting the onCampus parameter correctly. As I have explained, this parameter controls the level of access. We can set this parameter to true, if the user is a member of us (meaning that he has logged in with valid user credentials or that he is inside our network). But the institution code always is the same. However, with my PrimoPermissionHandler we are able to use the defaultCode and can set the defaultPermissionRule.
In a consortial environment we need to figure out the appropriate institution code. We can do that by setting the permissionRule['code'] = permissionRule. If one of the rules matches, we can always set the onCampus parameter to true. Thats the way the [Institution] option works currently. But this [Institution] has a failure: it always sets the oncampus parameter to true. We should not do that, if the user has matched only the global rule.

Though, if you say, that we can determine the home library with the permission system even now, I guess this PR is ready to be merged.

demiankatz commented:
The question I still have is, is there ever a situation where `permissionRule['code'] = permissionRule` will be used to determine 'code' but should not affect the 'onCampus' status? Should the determination of a code and the setting of the onCampus status be tied together, or are they independent axes?

Two possible proposals:

1.) If the existing system is adequate, maybe it would make sense to rename defaultPermissionRule to defaultOnCampusRule. Then we could include a comment like "If a specific permissionRule is matched above, the user will always be considered to be on campus. If the default code is used, the defaultOnCampusRule will be checked to determine whether they should be served full results (on campus) or limited public results (off campus)."

2.) If the existing system is not adequate, maybe we could change it like this:

insitutionCodeRule[inst1] = primo.isAtInst1
institutioncodeRule[inst2] = primo.isAtInst2
defaultInstitutionCode = inst3

onCampusRule[inst1] = primo.isOnCampusAtInst1
onCampusRule[inst2] = primo.isOnCampusAtInst2
defaultOnCampusRule = primo.isOnCampusAnywhere

For simplicity, we could allow boolean true or false to be used for onCampusRule settings to hard-code the on campus value to true or false without checking a permission. We could default the setting to "true" if unset since I suspect this will be the most common case... but then we have additional support for edge cases if we need it in the future.

I apologize if I am overthinking this -- it just seems like something that may cause more work down the road if we don't make it sufficiently flexible now. In any case, I'll let you use your judgment from here on, but I think I've presented the options as well as I can!

Thanks again for your help!

olli-gold commented:
Thank you for your input, Demian!

I have thought about your question and was about to say, that the system is adequate, because the only situation, where the code should be separate from onCampus access would be, that an active selection of the code would be possible. Than the Beluga consortium here in Hamburg came into my mind. There we have in deed exactly that situation: users can actively set the library they want, and this does not mean, that they are inside their network... Here we have the case you described: a code is independent from onCampus access...

In this scenario we could use your proposal 2 and set insitutionCodeRule[inst1] = primo.isAtInst1 for default code detection (Beluga has a switch, where users can select another library, and then another code needs to be used). This means, the default code detection needs to be overridable - if the user sets actively another library, this one needs to be used. Thats the point, where I think, that its getting obsolete again - because for autodetection we could also use onCampusRule[inst1] = primo.isOnCampusAtInst1 (and if nothing matches use the deafult code)... 

In general VuFind we do not have an active library selection, i.e. the Primo code can be determined the way this option works right now. So I think in the end its legitimate to couple code detection and onCampus access. For Beluga they need an additional mechanism to let the user switch the code. But this cannot be my work... I will ask the Beluga guys to review this here and to comment on this also. But this cannot be part of the configuration and needs other parameters.

However, I would prefer your proposal 1 and I think its a very good idea to rename the options, as this makes it more clear, what they mean.

Thanks again for your input, I will for the start implement your proposal number 1 and ask the Beluga guys for additional comments.

demiankatz commented:
Thanks, this sounds good. I'm definitely interested in hearing from Beluga, since I think they might end up advocating for proposal 2. However, another thought occurs to me for future consideration: it might be possible to implement multiple PermissionHandlers with different rules and then select the appropriate one based on the configuration. Obviously that's more work than we need to do right now, but in the future it might make sense to define a PrimoPermissionHandlerInterface and build a plugin manager from which different permission handler implementations could be retrieved based on different configuration settings.

EreMaijala commented:
I'm jumping right in the middle of this so please pardon me if I'm confused...

We run a consortium, but every institution has their own VuFind instance, so our case is pretty simple. What we need to be able to do is determine onCampus based on IP address or login. This would be simple but we have multiple login methods available and not all of them are authorizing. I'm not sure if what's proposed here would help us at all, but if it does affect the design, I'd appreciate it if would be taken into account (no need to provide support for checking authentication method but make the system extensible in that regard).

I'm not quite sure if  `defaultOnCampusRule = primo.isOnCampusAnywhere` makes sense. Since the search is always done in an institution regardless of the way the institution is chosen, onCampus is always relative to that institution.

olli-gold commented:
Thank you for your comment, Ere. 

You are right, defaultOnCampusRule = primo.isOnCampusAnywhere doesn't make sense. It has to be defaultOnCampusRule = primo.isOnDefaultCampus - but this is just a matter of the names. The defaultOnCampusRule is used to determine, which user has to be considered "on campus" for the default Primo institution code - nothing more.

"What we need to be able to do is determine onCampus based on IP address or login." -- Yes, that's exactly, why I have made this proposal, as this is not possible with the current PrimoCentral implementation in VuFind 2. The current implementation is looking for an IP match to determine the institution code and is setting onCampus always to true (even if the default institution code is getting used, this is normally not allowed). Though, this proposal changes that, making the assignation of the code more flexible (using the Permission system) and also making it possible to change the onCampus parameter for the default rule (for any other rule the parameter is always true).

I'm not sure, if that fits your needs, but I would appreciate if you could review the proposal and check if its useful for your scenario!

EreMaijala commented:
I believe this will be useful for us too (even if I'm struggling a bit in connecting permissions.ini directives with Primo.ini).

demiankatz commented:
I think Ere's example reinforces the idea that we need to decouple the insitution determination from the "on campus" permissioning. However, it's definitely true that my original "option 2" example did not make a lot of sense in its use of defaultOnCampusRule. Let me revise it to this:

insitutionCodeRule[inst1] = primo.isAtInst1
institutioncodeRule[inst2] = primo.isAtInst2
defaultInstitutionCode = inst3

onCampusRule[inst1] = primo.isOnCampusAtInst1
onCampusRule[inst2] = primo.isOnCampusAtInst2
onCampusRule[inst3] = primo.isOnCampusSomewhereElse

If we tie the on campus status to an institution code, there's actually no need at all for a defaultOnCampusRule... because the institution code rules will always give us some institution code, and then we just need a mapping from institution code to permission check. (And as I noted before, I think we should offer the option of boolean true/false here to force a permission or non-permission based on code, to make simple cases simpler).

olli-gold commented:
The proposal supplies basically three "Permission options":
onCampusRule[inst1] = primo.isOnCampusAtInst1
defaultOnCampusRule = primo.isOnDefaultCampus
defaultCode = default

This looks pretty similar to your proposal, doesn't it? But its a little bit simpler than your proposal, because it uses less options. The key question is: where is the difference between primo.isAtInst1 and primo.isOnCampusAtInst1? How could we determine the institution code automatically if not by using the primo.isOnCampusAtInst1 rule? 

I agree that it should be possible to set another institution code, but I believe this cannot be done with configuration. We would need a switch somewhere in the user interface for that. So, I think, the insitutionCodeRule is not usefull at all - we can use the onCampusRule for the same demand.

"If we tie the on campus status to an institution code, there's actually no need at all for a defaultOnCampusRule..." -- Why, yes! It might happen, that a user is on the default code and for that situation we definitely need the defaultOnCampusRule. For all other codes this is obsolete, that's why I would not use two rules for just the same thing.

The onCampus status must always be tied to a institution code, because we need to know, which user has full access to the results and which user hasn't. If the user wants to use another institution code (that's obviously only possible, if the institution has multiple institution codes, because VuFind is running in a consortium), we need to determine their permission rights for that institution code - and nothing else.

olli-gold commented:
I have to revise my last comment a little bit. Perhaps I understood you not completely correct.

You are right, perhaps we can drop the defaultOnCampusRule and use onCampusRule[default] = primo.isOnDefaultCampus. Right, this would simplify things a bit, and its easy to change. Is that waht you meant?

demiankatz commented:
Yes, that's exactly what I meant. We need three settings:

1.) A map that uses permissions to determine an institution code.
2.) A default institution code when the map finds no match.
3.) A map that translates institution codes into permissions we need to check to determine on campus permission.

I think that covers all use cases. Are you saying that's what you've already implemented? (I haven't reviewed the code in a while, so maybe I missed some changes... but I was under the impression we still had the simpler but less expressive options in place).

olli-gold commented:
Well, no, the defaultOnCampusRule is still a separate option. But that's pretty easy to change. Sorry, I think I had trouble understanding completely what you advised. 

I will simplify that. What do you think about the idea of dropping insitutionCodeRule[inst1] = primo.isAtInst1 also (leaving only onCampusRule[inst1] = primo.isOnCampusAtInst1)?

demiankatz commented:
If we drop institutionCodeRule, then we have no way of saying "this user belongs to institution 1, but is not currently on campus." It sounds like Ere is going to need that level of granularity. My proposal is that we use "institutionCodeRule" to figure out where the user belongs, and we use "onCampusRule[institutionCode]" to determine whether the user is on campus. If we default "onCampusRule" to boolean true in cases where it is undefined, then we can keep the configuration simple for the most common use cases while also offering a lot of flexibility. (Though it might be argued that defaulting a permission rule to true is dangerous, and the extra verbosity is worthwhile for added safety... I don't have a strong opinion here one way or another and would value additional input).

olli-gold commented:
'If we drop institutionCodeRule, then we have no way of saying "this user belongs to institution 1, but is not currently on campus."' -- Well, as I explained I think we cannot do that by configuration only. We need a switch somewhere in the user interface for that. By default, the default code (or the one mtching the onCampusRule) can get used, unless the user himself switches to another institution code or is set to another institutioncode by other means.

But perhaps Ere (or the Beluga guys) can say something about that.

demiankatz commented:
My argument is that we actually can do that by configuration only. We can define any arbitrary code as a permission handler, so regardless of what mechanism we use for selecting institution (user account attribute, UI control, or something else) there's a way to wire things up so that it causes a particular permission to be assigned. Obviously the details of that may or may not be complicated -- but from the perspective of the Primo code, I think we can safely say "set up a permission to provide this information" and rely on the user to make that work however they need to.

demiankatz commented:
(And, of course, if common patterns emerge for the way people do this, we can include tools to help make these tasks easier -- i.e. a standard way of creating a drop-down in the UI that turns a particular permission on might be useful).

olli-gold commented:
The latest commits are changing the defaultOnCampusRule to a normal onCampusRule for the defaultCode.
I haven't changed the function to autodetect a code yet (using separate options, not the onCampusRules), but you have convinced me, that this does make sense.

demiankatz commented:
Thanks! Let me know when that last step is done and I'll give this another close look.

olli-gold commented:
Ok, I have finished the last step and think, the thing is complete now and waiting for your (hopefully final) review.

demiankatz commented:
A couple more minor comments:

1.) It might be worth adding a PrimoBackendFactory::getPrimoPermissionHandler() method instead of creating it inline within createService(). By creating the support method, you make it easier for users to extend the factory and override the method to insert a custom permission handler.

2.) I think our ideas of how the settings should work are getting closer together, but they are not yet identical. My thought was that ONLY the institutionCode rule (in combination with defaultCode, of course) would be used to pick a code. THEN, onCampusRule would be used to determine on campus status for the rule that was picked. Your approach tries to use onCampusRule to determine institution code, which I think might cause problems in cases where criteria for code determination and on campus status overlap (for example, imagine you want to determine code based on a user attribute, but on campus based on an IP range -- then a user visiting a different campus would get the wrong code assigned).

You also haven't incorporated my suggestion of allowing onCampusRule to be boolean true or false to force an allow/deny setting without checking permissions (and that's fine, if you ignored it because you think it's a bad idea... but bringing it up in case it simply got missed).

But perhaps one idea that gives the fine control I'd like to see without forcing users to configure lots of extra settings would be this algorithm:

a.) Determine institution code using the institutionCode setting.
b.) Is an onCampusRule set for the selected institution code?
     i.) Yes, evaluate the permission rule specified (or pass boolean true/false through unchanged, if you want to support that possibility).
     ii.) No, just assume on campus (the assumption being that if no onCampusRule is set, then the onCampusRule is actually the same as the institutionCode rule)

Does that make any sense? Again, sorry to keep dragging this out -- but I want to be sure we're on the same page!

olli-gold commented:
I have added a getPermissionHandler methid to the PrimoBackendFactory. But I guess we still need to discuss about the second point.

> Your approach tries to use onCampusRule to determine institution code, which I think might cause problems in cases where criteria for code determination and on campus status overlap (for example, imagine you want to determine code based on a user attribute, but on campus based on an IP range -- then a user visiting a different campus would get the wrong code assigned).

Right, I am using institutionCode first, but when this does not match, the onCampus Rules are used to determine the institution code and onCampus access before switching to the default code. But I don't think this is a big problem, because its still possible to use both options (in cases, where code determination and onCampus status overlap). Doing that, I guess its a little bit more flexible as it allows users to skip the institutionCode settings, if onCampus settings are enough (and identical with code determination). However, I think, my solution is appropriate to combine easy handling and flexibility.

> for example, imagine you want to determine code based on a user attribute, but on campus based on an IP range -- then a user visiting a different campus would get the wrong code assigned

No, he wouldn't. The appropriate configuration for this case would be to set 
> institutionCode['MYINST'] = primo.isAtMyInst 
> institutionCode['OTHERINST'] = primo.isAtOtherInst 
> onCampusRule['MYINST'] = primo.isOnCampusAtMyInst
> onCampusRule['OTHERINST'] = primo.isOnCampusAtOtherInst

Assuming that primo.isAtMyInst matches, the code will be set to MYINST. Assuming it does not match (e.g. because the user did not log in and primo.isAtMyInst is using a user attribute to match), the code MYINST would still be applied, when the user is inside the network defined in the IP range of primo.isOnCampusAtMyInst. If the user is on campus of primo.isOnCampusAtOtherInst and the user did not log in, the code OTHERINST will get used (and thats correct). If the user logs in (still being member of MYINST), MYINST will be used, because primo.isAtMyInst matches.
This is slightly different the other way around (if primo.isAtMyInst and primo.isAtOtherInst are based on IP-ranges): OTHERINST will be applied, when the user is on campus of the other institution, but the user is not getting full access, because primo.isOnCampusAtOtherInst will not match. After the user logged in, he would still get the OTHERINST code (because primo.isAtOtherInst still matches and is overriding the code determination of the onCampusRule). Though, the instituion code needs to be set to MYINST on another way. 
This is what I think cannot be solved with configuration only, because in this case two rules match and will give different institution codes - how can we determine now, which one needs to be used? Currently we use the first match (the one from institutionCode), but the best way would be to use the code from the onCampusRule. This is not implemented right now, but your suggestion won't be able to determine this at all (as it would always use the institutionCode or the defaultCode).

> You also haven't incorporated my suggestion of allowing onCampusRule to be boolean true or false to force an allow/deny setting without checking permissions

You are right, I just missed that. But I also think, its not necessary at all, because we already have the hasPermission() method, which is doing a similar thing (and its also just returning a boolean value). I still don't understand, what would be the advantage of a boolean onCampusRule for that.

Did I convince you, that my proposal is good enough? Or do you still mean it would be better to seperate onCampusRules and institutionCodes more strictly?

demiankatz commented:
Thanks for the detailed feedback! I'm at a conference this week, but I'll try to get you a response as soon as I can. I just want to be sure I have enough time to read carefully and re-examine the code to make sure I'm understanding everything correctly before making a reply.

olli-gold commented:
All right, thanks for the information! Take your time, there's no need to hurry. And have a nice conference!

demiankatz commented:
I'm getting to this a little sooner than expected thanks to a generous lunch break!

I think you've convinced me at this point that we've come up with two separate models that probably are able to express more or less the same set of scenarios.

The main difference between our models is in how we represent the simplest possible case (you would omit institutionCodes, I would omit onCampusRules) and in how the underlying processing works (I think my approach more explicitly differentiates between the authorization vs. authentication aspects of the process, and might result in slightly less code... but it's not a huge difference). I still feel like your approach is a little more ambiguous in some way, but you've at least persuaded me that my original example is not adequate to break the model, so maybe it's not as much of a potential problem as my first gut feeling suggested.

At this point, I'd love to get a third opinion, since we're obviously both attached to our own models and some unbiased input might prove valuable. Perhaps @EreMaijala could be persuaded to comment (though I'd forgive him if he didn't feel like reading through this thread in depth).

In any case, I can certainly live with this as it is, though I still think it might be possible to improve upon it. I don't actually have to use or live with this code, though, so the importance of my opinion should be adjusted accordingly.

olli-gold commented:
Perfect, than lets wait for a third opinion first, either from Ere or from one of the Beluga guys (whom I have asked for comments, but this could take a little bit more time). 

Thank you again for your comments and suggestions!

EreMaijala commented:
Well, in our case there's no need to determine the institution code, just whether the user is authorized. So, for us it's enough if we can say that users authorized this way have access as well as users coming from these IP addresses. I might be confused (still), but I'm not sure how the rules for primo.isAtMyInst and primo.isOnCampusAtMyInst would differ.

olli-gold commented:
Tahnk you for your comment, Ere.

> I'm not sure how the rules for primo.isAtMyInst and primo.isOnCampusAtMyInst would differ

Yes, me also. I cannot imagine a case where this would make sense. But with the current design we are able to handle such cases, if there is a scenario, where it is needed, however.

I guess for Ere's scenario using one onCampusRule is enough. If access for all users to PrimoCentral is needed, a defaultCode can be added, otherwise the onCampusRule is enough.

demiankatz commented:
Okay, sounds like we can leave this alone, then. I'll give it one more thorough review upon returning from the conference, and then we should be able to merge. Thanks to both of you!

demiankatz commented:
Okay, I'm back from the conference. We're nearly there. Just a few more minor polishes:

1.) I don't think that [InstitutionOnCampus] is the best possible name for that config section. Would it actually make sense to merge this into the existing [Institutions] section and simply deprecate the old regex/code keys, or does that make it too hard to detect old vs. new configurations? If you think it's safer to keep the sections separate, how would you feel about changing [InstitutionOnCampus] to [InstitutionPermissions] or something else more generic. (My point is just that this section deals with more than just "on campus" settings -- it handles all aspects of institution authentication/authorization).

2.) The Primo example in permissions.ini doesn't really align with any of the examples in Primo.ini. Do you think we should change the permission name so at least that much matches, or do you think we should remove the example entirely, since it doesn't really demonstrate anything Primo specific? I'm fine with either approach -- but I think we should do something so we don't confuse people with inconsistent examples.

3.) I've taken another stab at revising the documentation in Primo.ini to make it more clear. Please take a look at what I've done and let me know if you think I've introduced any new problems.

4.) In the PrimoBackendFactory, the getPermissionHandler method isn't really a getter, since it's initializing a property and not returning anything. This is similar to a different change we made earlier -- I think the code would be easier to understand if this were a true getter. In fact, I think as written there's a bug -- createConnector() is relying on $this->primoPermissionHandler to set the institution code, but it is called PRIOR to getPermissionHandler(), so the property isn't initialized yet. If you just switched this to a getter and called the getter from createConnector, I think we'd be in better shape!

5.) I suspect that the PrimoPermissionHandler can actually be dramatically simplified. My understanding is that we essentially have two inputs here: a configuration, and an authentication service, which yield two outputs: an institution code, and an on-campus status. The permission handler handles some things that it might not need to -- do we really need things like getOnCampusRule and checkPermission, or could we just as easily set a flag during the autodetectCode process? If you think there's a need for all of this, we can certainly leave it alone -- but if we can slice out some complexity now, it might save us maintenance headaches down the road.

Let me know what you think! Obviously item 4 is the most important, since that looks to me like a significant bug. The rest is just designed to tidy things up, and I apologize if I'm being overly picky about any of it.

olli-gold commented:
Thank you for your new comments and for taking care about all of this!

1. I'm absolutely fine with renaming the section, but I would prefer not to merge it with the old settings to keep the old and the new section seperate as long as we support the old settings. I have noted that we had the name InstitutionPermissions earlier, and you are right, that it is really more appropriate for this.

2. I would prefer to leave a Primo-example in permissions.ini, but you are right, that it should be an appropriate example matching a default value from Primo.ini. I have corrected that.

3. Thank you, I guess you as a native speaker have a much better feeling how to describe things in english than me!

4. I have fixed the code according to your suggestions and hope, the bug has gone away with that.

5. I would prefer to leave the methods as they are, because they are usefull helper methods for the class. The autodetectCode process is only getting executed, if no institutionCode has been set (and it has not been tried to autodetect one yet). Though, if an instance is setting the institution code manually without relying on the autodetection (for example to force VuFind to use a certain code after a user has set a different code manually), these methods can help to determine the appropriate onCampusRule for that code and to check this rule.
I aggree, that the methods could be dropped, but than the contents would need to get integrated into other methods (checkPermission($code) into hasPermission() and getOnCampusRule into getDefaultOnCampusRule, but getOnCampusRule is also used in checkPermission and though this would probably duplicate code). I guess this would not really simplify things. But if you think this class is too complex I can spend some more time to try to simplify it.

demiankatz commented:
Thanks for the updates. This is looking much better now! Another round of comments (some new, some follow-ups to the previous round):

1. I'm still wondering if it's better to totally eliminate the [Institutions] section rather than keep it for backward compatibility. Since it's really easy to translate the existing section to the new section (just convert each pair into an institutionCode['CODE'] = primoPermission.CODE, and define primoPermission.CODE using the ipRegEx handler in permissions.ini) perhaps it would make sense to detect the presence of [Institutions] and throw an exception that says "add this to Primo.ini, add that to permissions.ini, and remove [Institutions]." Then we could remove the legacy code without dramatically inconveniencing the existing user base (since migration would be more or less automatic). I'd be happy to help with implementation of this if you think it's a good idea.

2. I made a few changes to InjectOnCampusListener to simplify the code and prevent the cache from persisting inappropriately if you swap out the permission handler. Tests still pass, so hopefully they're fine... but feel free to review if you like!

3. I've also done some work to shorten the PrimoPermissionHandler code (sorry if I got carried away). Tests are still passing, so I don't think I've broken anything -- but again, review wouldn't hurt. In any case, now that some verbosity and redundancy has been reduced, I feel better about the overall interface. Also, while reading the code, I noticed that I had misinterpreted something, which required me to fix my documentation (and also changed my opinion about the necessity of some of the methods).

Anyway, let me know what you think on item 1 (and whether I've messed anything up with my work on 2 and 3). We're very close to merging this, I'd say!

olli-gold commented:
Thanks again for the update and the new comments. 

Item 1 sounds good to me, and I would appreciate if you could help implementing that. I'll try to prepare something to help with it, but I'm sure you will be faster than me.

I have reviewed your modifications and tested them in our environment, and can confirm, that everything is working well. Perfect, thank you!

olli-gold commented:
Now I have restructured legacy support and merged the two sections into the old one (with the name Institutions).

I hope this spares you some time. If you have anything to add or to change, feel free to do it!

Thank you!

demiankatz commented:
Great, thanks! I'll work on the exception stuff later in the week!

demiankatz commented:
Hello again! I've just added Primo support to the VuFind\Config\Upgrade class, so the Primo settings can now be corrected as part of the automatic upgrade process. With this in place, it's probably safe to replace the logic in checkLegacySettings() with a simple `throw new \Exception('Legacy [Institutions] settings detected; please run upgrade process.);` (or something similar). What do you think?

olli-gold commented:
Hi, sorry for the delayed comment. I have replaced the legacy check with a simple exception now. I hope its all fine now and we can close this PR?

demiankatz commented:
Looks good! I've merged to master. Thanks for all the work!

olli-gold commented:
Great, thank you for merging!
